File: Interview Preparation Kit/Greedy Algorithms/LuckBalance.go
// https://www.hackerrank.com/challenges/luck-balance/problem
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
	"strings"
)

/*
 * Complete the 'luckBalance' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER k
 *  2. 2D_INTEGER_ARRAY contests
 */
func luckBalance(k int32, contests [][]int32) int32 {
	balance := int32(0)
	important := make([]int32, 0)

	for _, contest := range contests {
		if contest[1] == 0 {
			balance += contest[0]
		} else {
			important = append(important, contest[0])
		}
	}

	sort.Slice(important, func(i, j int) bool {
		return important[i] < important[j]
	})

	important_size := len(important)
	if important_size > int(k) {
		midpoint := important_size - int(k)
		for j := midpoint; j < important_size; j++ {
			balance += important[j]
		}

		for m := 0; m < midpoint; m++ {
			balance -= important[m]
		}
	} else {
		for _, val := range important {
			balance += val
		}
	}

	return balance
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	firstMultipleInput := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	nTemp, err := strconv.ParseInt(firstMultipleInput[0], 10, 64)
	checkError(err)
	n := int32(nTemp)

	kTemp, err := strconv.ParseInt(firstMultipleInput[1], 10, 64)
	checkError(err)
	k := int32(kTemp)

	var contests [][]int32
	for i := 0; i < int(n); i++ {
		contestsRowTemp := strings.Split(strings.TrimRight(readLine(reader), " \t\r\n"), " ")

		var contestsRow []int32
		for _, contestsRowItem := range contestsRowTemp {
			contestsItemTemp, err := strconv.ParseInt(contestsRowItem, 10, 64)
			checkError(err)
			contestsItem := int32(contestsItemTemp)
			contestsRow = append(contestsRow, contestsItem)
		}

		if len(contestsRow) != 2 {
			panic("Bad input")
		}

		contests = append(contests, contestsRow)
	}

	result := luckBalance(k, contests)

	fmt.Fprintf(writer, "%d\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}

File: Interview Preparation Kit/Greedy Algorithms/LuckBalance.js
// https://www.hackerrank.com/challenges/luck-balance/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the luckBalance function below.
function luckBalance(k, contests) {
    let balance = 0;
    let important = [];

    for (let i = 0; i < contests.length; i++) {
        if (contests[i][1] == 0) {
            balance += contests[i][0];
        }
        else {
            important.push(contests[i][0]);
        }
    }

    important.sort(function(a, b) {
        return a - b;
    });

    if (important.length > k) {
        let midpoint = important.length - k;
        for (let j = midpoint; j < important.length; j++) {
            balance += important[j];
        }

        for (let m = 0; m < midpoint; m++) {
            balance -= important[m];
        }
    }
    else {
        for (let p = 0; p < important.length; p++) {
            balance += important[p];
        }
    }

    return balance;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const nk = readLine().split(' ');

    const n = parseInt(nk[0], 10);

    const k = parseInt(nk[1], 10);

    let contests = Array(n);

    for (let i = 0; i < n; i++) {
        contests[i] = readLine().split(' ').map(contestsTemp => parseInt(contestsTemp, 10));
    }

    const result = luckBalance(k, contests);

    ws.write(result + '\n');

    ws.end();
}
